### ISSUE TITLE
TypeScript issue when passing Pino instance to fastify constructor

### ISSUE THREAD

**Issue by pavelkornev** [fastify/fastify#4960]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Fastify version

4.21.0

### Node.js version

18.14.2

### Operating system

macOS

### Operating system version (i.e. 20.04, 11.3, 10)

13.5

### Description

By some historical reasons we pass a Pino instance to fastify factory function (I'm aware that Pino is built-in into fastify):

```
const logger = pino({
  // ...some config
});

const server = fastify({
  logger,
  // ...other opts
});

```

But starting `v4.20.0` this code is no longer working with the following TypeScript compile error:

```
Type 'FastifyInstance<RawServerDefault, IncomingMessage, ServerResponse<IncomingMessage>, Logger<LoggerOptions | DestinationStream>, FastifyTypeProviderDefault>' is not assignable to type 'FastifyInstance<RawServerDefault, IncomingMessage, ServerResponse<IncomingMessage>, FastifyBaseLogger, FastifyTypeProviderDefault>'.
  The types of 'withTypeProvider().childLoggerFactory' are incompatible between these types.
    Type 'FastifyChildLoggerFactory<RawServerDefault, IncomingMessage, ServerResponse<IncomingMessage>, Logger<LoggerOptions | DestinationStream>, any>' is not assignable to type 'FastifyChildLoggerFactory<RawServerDefault, IncomingMessage, ServerResponse<IncomingMessage>, FastifyBaseLogger, any>'.
      The 'this' types of each signature are incompatible.
        Type 'FastifyInstance<RawServerDefault, IncomingMessage, ServerResponse<IncomingMessage>, FastifyBaseLogger, any>' is not assignable to type 'FastifyInstance<RawServerDefault, IncomingMessage, ServerResponse<IncomingMessage>, Logger<LoggerOptions | DestinationStream>, any>'.
          Types of property 'log' are incompatible.
            Type 'FastifyBaseLogger' is not assignable to type 'Logger<LoggerOptions | DestinationStream>
```

It seems to be related to this recent change - https://github.com/fastify/fastify/pull/4760

### Steps to Reproduce

1. Create Pino instance
2. Try to pass to `fastify` factory function

**Comment by mcollina** [comment]
Thanks for reporting!

Can you provide steps to reproduce? We often need a [reproducible example](https://stackoverflow.com/help/minimal-reproducible-example), e.g. some code that allows someone else to recreate your problem by just copying and pasting it. If it involves more than a couple of different file, create a new repository on GitHub and add a link to that.

**Comment by ChrisLahaye** [comment]
We have the same issue but its because we are assigning the fastify instance to a variable with type FastifyInstance. Since we don't specify the generics the logger type defaults to FastifyBaseLogger, which doesn't seem to be compatible with any pino instance. Seems like the new FastifyBaseLogger type is not correct.

**Comment by pavelkornev** [comment]
> We have the same issue but its because we are assigning the fastify instance to a variable with type FastifyInstance. Since we don't specify the generics the logger type defaults to FastifyBaseLogger, which doesn't seem to be compatible with any pino instance. Seems like the new FastifyBaseLogger type is not correct.

Yeah, we also assign the server to a variable of type `FastifyInstance` and it fails starting `v4.20.0`.

**Comment by ZiiMakc** [comment]
@mcollina this is example:
```
import { fastify, type FastifyInstance } from "fastify";
import { pino } from "pino";

// all latest versions
const log2 = pino();
const app2: FastifyInstance = fastify({
	logger: log2,
});
```

Maybe related to https://github.com/fastify/fastify/pull/4928

**Comment by Uzlopak** [comment]
Why should it be related #4928?

**Comment by mcollina** [comment]
@ZiiMakc I've literally tested your example with the top-most version of Fastify, Pino and TypeScript and I could not reproduce. Please provide a full repro including the `tsconfig.json` and all relevant assets.

**Comment by ZiiMakc** [comment]
@mcollina here it is: https://github.com/ZiiMakc/fastify-type-errr

run:
```
pnpm i
pnpm lint
```

**Comment by mcollina** [comment]
This is confirmed only with `"strict": true` in the `tsconfig.json`.

There seem to be an issue with:

```
Type 'FastifyBaseLogger' is not assignable to type 'Logger<LoggerOptions | DestinationStream>'
```

Which I think it's due to:

https://github.com/fastify/fastify/blob/c235d2dc136a67c51095f188742c743d8cb32839/types/logger.d.ts#L24

not being compatible with

https://github.com/pinojs/pino/blob/91a3505cfaef323f565e3cfe93f5d5077320faff/pino.d.ts#L83

I don't know why this problem shows up only in strict mode and how to change this. At a first glance these seems very similar. However some of those generics are slightly different.

Anyhow, this fixes your problem:

```ts
import { fastify, type FastifyInstance, type FastifyBaseLogger } from "fastify";
import { pino } from "pino";

// all latest versions
const log = pino();
const app: FastifyInstance = fastify({
  logger: log as FastifyBaseLogger,
});
```

**Comment by rdzidziguri** [comment]
We faced the same problem, and the only solution is to use the `as` keyword which is not ideal, but it does the job for now.

**Comment by NotEvenANeko** [comment]
In the

```typescript
const log = pino();
const app: FastifyInstance = fastify({
  logger: log,
});

```

case, we are assigning `FastifyInstance<..., pino.Logger, ...>` to `FastifyInstance<..., FastifyBaseLogger, ...>`, so the type of `FastifyInstance<..., pino.Logger, ...>.childLoggerFactory` (or some other functions) need to be compatible with `FastifyInstance<..., FastifyBaseLogger, ...>.childLoggerFactory`.

And the `childLoggerFactory` function takes a `this` parameter as type `FastifyInstance`, so it means we need to assign a `(this: FastifyInstance<..., pino.Logger, ...>): ...` to `(this: FastifyInstance<..., FastifyBaseLogger, ...>): ...`.

And function parameters in TypeScript are **contravariant** when `strict: true` or `strictFunctionTypes: true`, which means even though `pino.Logger` is the **subtype** of `FastifyBaseLogger` (`FastifyBaseLogger` is `pino.BaseLogger` with a `child` function), the function `(this: FastifyInstance<..., pino.Logger, ...>): ...` is the **basetype** of `(this: FastifyInstance<..., FastifyBaseLogger, ...>): ...`, and we can't assign a basetype to a subtype.

https://www.typescriptlang.org/docs/handbook/type-compatibility.html#comparing-two-functions

This should also work, `app` should be `FastifyInstance<..., pino.Logger, ...>` now.

```typescript
const log = pino();
const app = fastify({
  logger: log,
});
```

**Comment by mcollina** [comment]
How would you fix it?

**Comment by mcollina** [comment]
@flakey5 could you take a look?

**Comment by flakey5** [comment]
From my testing I believe I can second what @NotEvenANeko said about this.

It's not so much `FastifyBaseLogger` being incompatible with `Logger<LoggerOptions | DestinationStream>`, this works without error
```typescript
const log = pino();
const a: FastifyBaseLogger = log;
const b: boolean | FastifyLoggerOptions<RawServerDefault> & PinoLoggerOptions | Logger = log; // type of the `logger` property in `FastifyOptions`
```
but instead `pino.Logger` becoming the base type rather than `FastifyBaseLogger`. So for example a function `abc(logger: Logger)` needs either an instance of a `pino.Logger` or something that extends `pino.Logger`.

It does appear to be connected to #4760, removing `childLoggerFactory` from the `FastifyInstance` type fixes the problem

Few things we can do to fix it (starting with good ending with bad),

 1. Swapping this line https://github.com/fastify/fastify/blob/9ab24be4621e4988e682686843664a42f4071f3d/types/instance.d.ts#L484 for `childLoggerFactory: FastifyChildLoggerFactory<RawServer, RawRequest, RawReply, FastifyBaseLogger, TypeProvider>;` - _Should_ be good, but it does throw off the precedent of just passing in `Logger`
 2. Replace this line https://github.com/fastify/fastify/blob/9ab24be4621e4988e682686843664a42f4071f3d/types/instance.d.ts#L123 with `Logger = any` - Not great, decreased type safety and might be breaking
 3. Remove `Logger` from `FastifyInstance`'s generic fields and always use `FastifyBaseLogger` - Not great, definitely breaking

**Comment by mcollina** [comment]
@flakey5 would you like to send a PR for option 1?

**Comment by NikhilDendeti** [comment]
I'd like to help with this. Given it's traced to the childLoggerFactory typing introduced in #4760, this is likely a case where FastifyBaseLogger's type needs to be made compatible with (or genericized over) the Logger<LoggerOptions | DestinationStream> type from Pino, rather than assuming a specific logger shape. I'll reproduce the type error locally and check whether relaxing/adjusting the childLoggerFactory signature fixes it without breaking other typing tests.

**Comment by macrogui** [comment]
Picking this up with #7035 — it implements the "option 1" direction agreed above (typing the `childLoggerFactory` member against `FastifyBaseLogger`, extended to the `RouteShorthandOptions` override that kept the invariance alive through `route()`).

Verified the fix against current `main` (v6.0.0-alpha.4): `fastify({ loggerInstance: pino() })` still cannot be assigned to a default `FastifyInstance` today, and the new regression assertions fail without the type change. @NikhilDendeti FYI since you offered to help here in July — happy to collaborate if you were mid-way through something.
